File: SDE_sheet/Linked-List/07_intersection/solution.rb
# Definition for singly-linked list.
# class ListNode
#     attr_accessor :val, :next
#     def initialize(val)
#         @val = val
#         @next = nil
#     end
# end

# @param {ListNode} headA
# @param {ListNode} headB
# @return {ListNode}
def getIntersectionNode(headA, headB)
    a = headA
    b = headB
    while a != b
      a = a ? a.next : headB
      b = b ? b.next : headA
    end
    a
end

File: SDE_sheet/Linked-List/07_intersection/solution.php
/**
 * Definition for a singly-linked list.
 * class ListNode {
 *     public $val = 0;
 *     public $next = null;
 *     function __construct($val) { $this->val = $val; }
 * }
 */

class Solution {
    /**
     * @param ListNode $headA
     * @param ListNode $headB
     * @return ListNode
     */
    function getIntersectionNode($headA, $headB) {
        $a = $headA;
        $b = $headB;
        while ($a !== $b) {
            $a = $a !== null ? $a->next : $headB;
            $b = $b !== null ? $b->next : $headA;
        }
        return $a;
    }
}
